heroku/cli
# fix(pg-v5): Properly close ssh connection

https://heroku.support/939668
[W-8530613](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008kgP4IAI/view)

In the case that the following events happen:

* `heroku pg:psql` is run on a **private-tier** (not shield) postgres plan  in a **private space** 
* Connection to postgres server times out (leave terminal open for 10-15 minutes to reproduce this)
* When psql reconnects due to timeout
* And user exits program (ctrl+D or types `exit` or `\q` into psql terminal)
* psql exits successfully

Then the heroku CLI process should exit. Before this commit, the process would hang because the tunnel created by the `tunnel-ssh` package does not seem to close itself in this scenario of reconnection. This PR does a bit of refactoring so we can retain a reference to the `tunnel-ssh` server instance, and always call `.close` ourselves when psql exits. Or kill psql ourselves when the server is closed for whatever reason.


## Other stuff included

### ctrl-c handler

I also brought back the Ctrl-C handler removed in https://github.com/heroku/cli/pull/1687; turns out that psql swallows ctrl+c (you have to use ctrl+d to kill psql). This is useful and important to have so that you can kill a long-running query while running psql. I've restored it with a comment that hopefully makes it a bit more clear why the code was necessary. We need to swallow the ctrl-c from Node so that the node process isn't killed (if the parent node process is killed it will take the child psql process, and the SSH tunnel, with it)

### test refactoring

The tests contained a lot of duplication, so I created a new `mockSpawn` helper to help with that. I also created some fake child process stubs and `ssh-tunnel` stubs so we could better test the failure cases this PR was trying to address.


## How to test?

* Pull down this branch
* Run `cd packages/pg-v5 && heroku plugins:link`
* Run `heroku psql -a chad-test-space-app HEROKU_POSTGRESQL_ORANGE`
* psql should connect. You should be able to run queries.
* Leave the terminal psql console open for 15 minutes
* type a query like `select now();` into the psql console and hit enter
* You should see a message about disconnecting and reconnecting over SSL from psql
* quit psql with ctrl+d or by typing `exit` or `\q` into psql and hitting enter
* the CLI process should quit
* **Be sure to run `heroku plugins:unlink` after testing**
